ENH: Add BubbleSAM detection method

nidhinthomas-ai opened the pull request:
This merge request introduces a new library module, `neat_ml/bubblesam`, which wraps the `Segment Anything‑2 (SAM‑2)` model for detecting circular “bubble” features in microscopy images. It also adds an accompanying test suite with deterministic baseline image comparisons.

**What’s inside**

* **BubbleSAM Blob Detection**
  * `neat_ml/bubblesam/SAM.py`: Wrapper around SAM‑2 providing:
    * model construction via build_sam2(...) using a YAML config + checkpoint.
    * setup_cuda() helper to enable bfloat16/TF32 on Ampere‑class GPUs for speed.
    * generate_masks(...) to run SAM2AutomaticMaskGenerator on an RGB image and return masks sorted by area.
  * neat_ml/bubblesam/bubblesam.py End‑to‑end bubble detection utilities built on SAMModel
    * analyze_and_filter_masks(...) computes shape metrics (area, circularity, axes)
    * plot_filtered_masks(...) draws blue contours and green bounding boxes for retained regions.
    * process_image(...) runs the full SAM‑2 -\> filter -\> (optional) debug overlay/contours flow and persists artifacts/summary.
    * run_bubblesam(...) batch‑orchestrator returning one‑row‑per‑image summaries with median_radii_SAM and num_blobs_SAM.
* **Running bubblesam detection**
Instructions for running `bubblesam` detection have been added to the `README.md` file

Commit:
ENH: Add scripts for OpenCV detection

Add preprocessing.py for enhancing input images
Add detection.py for OpenCV blob detection
Add input data for testing OpenCV detection
Add scripts for testing OpenCV detection
Add raw_processed.png to baseline for test
Add raw_detection.png to baseline for test
Modified pyproject.toml to pull .tiff and .png

Commit:
TST: Modified test scripts

Removed fixtures from test scripts
Use importlib.resources to retrieve images
Reduced the tolerances for compare_images()
Added ruff.toml to enforce no mutable default args

Commit:
ENH: Added BubbleSAM detection method

Added bubblesam.py and SAM.py into neat_ml/bubblesam
Added stage_bubblesam() into run_workflow.py
Added test_bubblesam.py and associated baseline images
Modified .gitlab-ci.yml, mypy.ini and pyproject.toml
to incorporate sam2 module
Updated README.md with sam-2 installation instructions

Added bubblesam_detection_test.yaml for functional test

Commit:
MAINT: Added .gitmodules for sam2

Commit:
TST: Added baseline images for the bubblesam test

Commit:
MAINT: Add sam2 as submodule

Commit:
MAINT: Add sam2 as submodule

Commit:
DOC: Updated README.md with instructions

Commit:
ENH: Created workflow module

Created neat_ml/workflow and added lib_workflow.py
Added test_workflow.py to test lib_workflow.py

Commit:
CI: Added torch module to .gitlab-ci.yml

Commit:
TST: Modified test_bubblesam.py to improve coverage

Commit:
CI: Added .github/workflows/ci.yml

Added .github/workflows/ci.yml
Added LANL copyright assertion ID

adamwitmer commented:
Initial TODO items for reviewing this branch:

- [ ] read/understand diff/PR
- [ ] rebase branch against `main` (and push backup branch)
- [ ] fix issues with `git lfs` (i.e. image storage with `zenodo`)
- [ ] run test-suite and check test coverage 
- [ ] run branch according to `README.md` instructions
- [ ] copy remaining review comments from `gitlab` 
- [ ] perform detailed code review
- [ ] address all review comments

adamwitmer commented on README.md:
```diff
@@ -1,11 +1,68 @@
+conda activate ldrd_neat_ml
+
+conda install pytorch==2.5.1 torchvision==0.20.1 torchaudio==2.5.1 pytorch-cuda=12.4 -c pytorch -c nvidia
+pip install -r requirements.txt
```
from `gitlab`:

> typically one would want to avoid mixing pip and conda, by providing i.e., separate instructions/configs for each

adamwitmer commented on neat_ml/bubblesam/SAM.py:
```diff
@@ -0,0 +1,144 @@
+from pathlib import Path
+from typing import Any, Dict, List, Optional
```
change all instances of `Dict` and `List` to their native python types, i.e. `dict`, `list` here and elsewhere

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,330 @@
+
+    Returns
+    -------
+    pd.DataFrame
```
`df_filtered : pd.DataFrame`

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,330 @@
+
+    Returns
+    -------
+    pd.DataFrame
```
`filtered_df : pd.DataFrame`

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,330 @@
+        One row per image with detection statistics.
+    """
+    cfg_m = {**DEFAULT_MODEL_CFG, **(model_cfg or {})}
+    cfg_s = {**DEFAULT_MASK_SETTINGS, **(mask_settings or {})}
```
these are being used as module global instead of explicitly passed arguments

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,330 @@
+    summary.fillna(0, inplace=True)
+
+    (out_dir / "bubblesam_summary.csv").write_text(summary.to_csv(index=False))
+    print(f"[BubbleSAM] processed {len(df_imgs)} images -> {out_dir}")
```
should be switch to use `logging`

adamwitmer commented on neat_ml/opencv/preprocessing.py:
file needs to be rebased after merge of PR #2

adamwitmer commented on neat_ml/bubblesam/SAM.py:
```diff
@@ -0,0 +1,144 @@
+
+        Returns
+        -------
+        List[Dict[str, Any]]
```
`masks_sorted : `

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,330 @@
+import pandas as pd
+import matplotlib.pyplot as plt
+import cv2
+from typing import List, Dict, Any, Optional, Union
```
here and below replace instance of `Dict` and `List` with `dict` and `list`

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,330 @@
+    Parameters
+    ----------
+    image_path : str
+                 The path to the image file.
```
awkward spacing here.

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,330 @@
+    Parameters
+    ----------
+    df_imgs : pd.DataFrame
+        Requires 'image_filepath'.
```
here and below, these docstrings are not very descriptive of the actual function arguments.

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,220 @@
+    fig, ax = plt.subplots()
+    plt.sca(ax)
+    show_anns(masks)
+    plt.close(fig)
```
Im not really sure what the purpose of this test is...it doesnt have any assertions not does it test any errors/warnings, etc. Perhaps there is a way to design a more rigorous test for this function.

adamwitmer commented on neat_ml/tests/test_detection.py:
```diff
@@ -0,0 +1,82 @@
+        actual_png  = actual_dir/f"{stem}_debug.png"
+        desired_png = baseline_dir/f"{stem}_detection.png"
+
+        compare_images(str(desired_png), str(actual_png), tol=1e-4)
```
`compare_images` needs assertion

adamwitmer commented on neat_ml/tests/test_preprocessing.py:
These tests will be rebased against PR #2

adamwitmer commented on neat_ml/workflow/lib_workflow.py:
```diff
@@ -0,0 +1,163 @@
+    df_imgs = build_df_from_img_paths(img_paths)
+    run_opencv(df_imgs, det_dir, debug=debug)
+
+def stage_bubblesam(dataset_config: Dict[str, Any], paths: Dict[str, Path]) -> None:
```
It may be possible to _carefully_ combine `stage_opencv` with `stage_bubblesam` by passing `method` to a single function that has the capability to run both.

adamwitmer commented on neat_ml/bubblesam/SAM.py:
```diff
@@ -0,0 +1,144 @@
+
+        Returns
+        -------
+        None
```
purge `Returns: None`

adamwitmer commented on neat_ml/bubblesam/SAM.py:
```diff
@@ -0,0 +1,144 @@
+        Parameters;
+                none
+        Returns
+                None
```
purge `Parameters: None` and `Returns: None`

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,220 @@
+    """
+    return SAMModel(
+        model_config="sam2_hiera_l.yaml",
+        checkpoint_path=str(CHECKPOINT),
```
checkpoint also being passed as module global here.

adamwitmer commented on .github/workflows/ci.yml:
this needs to be carefully rebased to include the installation steps for `SAM2`

adamwitmer commented on pyproject.toml:
need to add `torch` to `pyproject` dependencies after rebase

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,220 @@
+    expected_cols = {"image_filepath", "median_radii_SAM", "num_blobs_SAM"}
+    npt.assert_(expected_cols.issubset(set(summary.columns)), "Missing expected columns")
+    arr: NDArray[np.int64] = np.asarray(summary["num_blobs_SAM"], dtype=np.int64)
+    npt.assert_array_less(0, arr)
```
perhaps other assertions can be made about the return `summary`, however.

ajwitmer committed:
MAINT: address reviewer comments

* remove `.gitattributes` and `.gitlab-ci.yml`
* revise `README.md`
* remove `Dict`, `List`, `Tuple` typing
* fix docstrings
* add inline comment explaining selection of bubblesam params with reference to issue #13
* combine duplicate functions and parameterize tests
* fix test assertions

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,349 @@
+        cv2.imwrite(os.path.join(output_path, f'mask_{i}.png'), mask_image)
+
+
+def show_anns(anns: list[dict[str, Any]]) -> None:
```
Also potentially could just remove this function in favor of saving mask overlays with `debug` flag

adamwitmer replied:
We can add a caveat here expressing that this code was derived from the above source.

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,349 @@
+        circ = (4.0 * np.pi * area) / (perimeter ** 2)
+        major_axis = rp.major_axis_length
+        minor_axis = rp.minor_axis_length
+        if area >= area_threshold and circ >= circularity_threshold:
```
I considered that the logic used in this function to check for `area` and `circularity` thresholds before indexing the dataframe might be improved by allocating an empty dataframe and indexing all values before performing filtering. However, the dictionary assignment performs `find_contours` on the segmentation map, which takes up a significant portion (21.4%) of the overall function call time. This can be seen be using `line_profiler`:

<details><summary>original logic (2.42 sec)</summary>

```bash
Timer unit: 1e-06 s

Total time: 2.41901 s
File: /Users/awitmer/Documents/LANL/2025/github_projects/ldrd_neat_ml/neat_ml/bubblesam/bubblesam.py
Function: analyze_and_filter_masks at line 116

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   116                                           @line_profiler.profile
   117                                           def analyze_and_filter_masks(
   118                                               masks_summary_df: pd.DataFrame,
   119                                               area_threshold: float = 25,
   120                                               circularity_threshold: float = 0.9
   121                                           ) -> pd.DataFrame:
   122                                               """
   123                                               Analyzes each mask (row) in masks_summary_df, measuring shape properties 
   124                                               and filtering out non-circular masks based on a circularity threshold.
   125                                           
   126                                               Parameters
   127                                               ----------
   128                                               masks_summary_df : pd.DataFrame
   129                                                   DataFrame containing at least a column 'segmentation' with boolean masks.
   130                                               area_threshold : float
   131                                                   Minimum area for the mask to be retained. Value of minimum area was determined
   132                                                   by hand-tuning based on visual observation to exclude objects that were
   133                                                   too small to be considered bubbles in the microscopy images 
   134                                               circularity_threshold : float
   135                                                   Circularity threshold to consider for the mask to be "circular."
   136                                                   Value for threshold was determined by hand-tuning based on visual inspection
   137                                                   to exclude "debris" particles while accounting for non-perfectly circular
   138                                                   shape of bubbles in microscopy images
   139                                                   
   140                                               Returns
   141                                               -------
   142                                               df_filtered : pd.DataFrame
   143                                                   A filtered DataFrame with new columns (contour, bbox, major_axis, minor_axis, 
   144                                                   area, radius, circ, euler_number) but excluding 'segmentation'.
   145                                               """
   146         2          1.0      0.5      0.0      filtered_rows = []
   147                                           
   148        56       6540.0    116.8      0.3      for idx, row in masks_summary_df.iterrows():
   149        54       1018.0     18.9      0.0          seg = row['segmentation']
   150        54    1009341.0  18691.5     41.7          labeled_seg = label(seg)
   151        54     581774.0  10773.6     24.1          props_list = regionprops(labeled_seg)
   152        54         22.0      0.4      0.0          if len(props_list) == 0:
   153                                                       continue
   154                                           
   155        54      11373.0    210.6      0.5          rp = props_list[0]
   156        54       4356.0     80.7      0.2          area = rp.area
   157        54     186784.0   3459.0      7.7          perimeter = rp.perimeter
   158        54         25.0      0.5      0.0          if perimeter == 0:
   159                                                       continue
   160                                           
   161        54         53.0      1.0      0.0          circ = (4.0 * np.pi * area) / (perimeter ** 2)
   162        54      94704.0   1753.8      3.9          major_axis = rp.major_axis_length
   163        54        465.0      8.6      0.0          minor_axis = rp.minor_axis_length
   164        54         17.0      0.3      0.0          if area >= area_threshold and circ >= circularity_threshold:
   165        29         23.0      0.8      0.0              mask_info = {              
   166        29        101.0      3.5      0.0                  'bbox': rp.bbox,
   167        29     516507.0  17810.6     21.4                  'contour': find_contours(seg, level=0.5)[0],
   168        29          9.0      0.3      0.0                  'major_axis': major_axis,
   169        29          3.0      0.1      0.0                  'minor_axis': minor_axis,
   170        29          7.0      0.2      0.0                  'area': area,
   171        29         90.0      3.1      0.0                  'radius': np.sqrt(area / np.pi),
   172        29          4.0      0.1      0.0                  'circ': circ,
   173        29       4495.0    155.0      0.2                  'euler_number': rp.euler_number
   174                                                       }
   175                                           
   176        29         13.0      0.4      0.0              filtered_rows.append(mask_info)
   177                                           
   178         2       1279.0    639.5      0.1      df_filtered = pd.DataFrame(filtered_rows)
   179         2          2.0      1.0      0.0      return df_filtered

  2.42 seconds - /Users/awitmer/Documents/LANL/2025/github_projects/ldrd_neat_ml/neat_ml/bubblesam/bubblesam.py:116 - analyze_and_filter_masks
```

</details>

Whereas with the following diff, the function runs slower because it performs `find_contours` on every segmentation map (32.1%):

<details><summary>Diff</summary>

```diff
 def analyze_and_filter_masks(
     masks_summary_df: pd.DataFrame,
     area_threshold: float = 25,
@@ -143,8 +143,10 @@ def analyze_and_filter_masks(
         A filtered DataFrame with new columns (contour, bbox, major_axis, minor_axis, 
         area, radius, circ, euler_number) but excluding 'segmentation'.
     """
-    filtered_rows = []
-
+    filtered_rows = pd.DataFrame(index=range(len(masks_summary_df)),
+        columns=["bbox", "contour", "major_axis", "minor_axis", "area",
+            "radius", "circ", "euler_number"]).fillna(np.nan)
+    filtered_rows[["bbox", "contour"]] = filtered_rows[["bbox", "contour"]].astype('object')  
     for idx, row in masks_summary_df.iterrows():
         seg = row['segmentation']
         labeled_seg = label(seg)
@@ -161,21 +163,22 @@ def analyze_and_filter_masks(
         circ = (4.0 * np.pi * area) / (perimeter ** 2)
         major_axis = rp.major_axis_length
         minor_axis = rp.minor_axis_length
-        if area >= area_threshold and circ >= circularity_threshold:
-            mask_info = {              
-                'bbox': rp.bbox,
-                'contour': find_contours(seg, level=0.5)[0],
-                'major_axis': major_axis,
-                'minor_axis': minor_axis,
-                'area': area,
-                'radius': np.sqrt(area / np.pi),
-                'circ': circ,
-                'euler_number': rp.euler_number
-            }
-
-            filtered_rows.append(mask_info)
-
-    df_filtered = pd.DataFrame(filtered_rows)
+        mask_info = {              
+            'bbox': rp.bbox,
+            'contour': find_contours(seg, level=0.5)[0],
+            'major_axis': major_axis,
+            'minor_axis': minor_axis,
+            'area': area,
+            'radius': np.sqrt(area / np.pi),
+            'circ': circ,
+            'euler_number': rp.euler_number
+        }
+
+        filtered_rows.iloc[idx] = pd.Series(mask_info)
+
+    df_filtered = filtered_rows[(filtered_rows["area"] > area_threshold) &
+        (filtered_rows["circ"] >= circularity_threshold)]
     return df_filtered
```

</details>

<details><summary>With Diff (3.13 sec)</summary>

```bash
Timer unit: 1e-06 s

Total time: 3.13041 s
File: /Users/awitmer/Documents/LANL/2025/github_projects/ldrd_neat_ml/neat_ml/bubblesam/bubblesam.py
Function: analyze_and_filter_masks at line 116

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   116                                           @line_profiler.profile
   117                                           def analyze_and_filter_masks(
   118                                               masks_summary_df: pd.DataFrame,
   119                                               area_threshold: float = 25,
   120                                               circularity_threshold: float = 0.9
   121                                           ) -> pd.DataFrame:
   122                                               """
   123                                               Analyzes each mask (row) in masks_summary_df, measuring shape properties 
   124                                               and filtering out non-circular masks based on a circularity threshold.
   125                                           
   126                                               Parameters
   127                                               ----------
   128                                               masks_summary_df : pd.DataFrame
   129                                                   DataFrame containing at least a column 'segmentation' with boolean masks.
   130                                               area_threshold : float
   131                                                   Minimum area for the mask to be retained. Value of minimum area was determined
   132                                                   by hand-tuning based on visual observation to exclude objects that were
   133                                                   too small to be considered bubbles in the microscopy images 
   134                                               circularity_threshold : float
   135                                                   Circularity threshold to consider for the mask to be "circular."
   136                                                   Value for threshold was determined by hand-tuning based on visual inspection
   137                                                   to exclude "debris" particles while accounting for non-perfectly circular
   138                                                   shape of bubbles in microscopy images
   139                                                   
   140                                               Returns
   141                                               -------
   142                                               df_filtered : pd.DataFrame
   143                                                   A filtered DataFrame with new columns (contour, bbox, major_axis, minor_axis, 
   144                                                   area, radius, circ, euler_number) but excluding 'segmentation'.
   145                                               """
   146         6       3649.0    608.2      0.1      filtered_rows = pd.DataFrame(index=range(len(masks_summary_df)),
   147         2          3.0      1.5      0.0          columns=["bbox", "contour", "major_axis", "minor_axis", "area",
   148         2       1163.0    581.5      0.0              "radius", "circ", "euler_number"]).fillna(np.nan)
   149         2       3017.0   1508.5      0.1      filtered_rows[["bbox", "contour"]] = filtered_rows[["bbox", "contour"]].astype('object')  
   150        56       4641.0     82.9      0.1      for idx, row in masks_summary_df.iterrows():
   151        54        744.0     13.8      0.0          seg = row['segmentation']
   152        54    1019170.0  18873.5     32.6          labeled_seg = label(seg)
   153        54     589941.0  10924.8     18.8          props_list = regionprops(labeled_seg)
   154        54         22.0      0.4      0.0          if len(props_list) == 0:
   155                                                       continue
   156                                           
   157        54      14066.0    260.5      0.4          rp = props_list[0]
   158        54       4071.0     75.4      0.1          area = rp.area
   159        54     185410.0   3433.5      5.9          perimeter = rp.perimeter
   160        54         22.0      0.4      0.0          if perimeter == 0:
   161                                                       continue
   162                                           
   163        54         64.0      1.2      0.0          circ = (4.0 * np.pi * area) / (perimeter ** 2)
   164        54      96108.0   1779.8      3.1          major_axis = rp.major_axis_length
   165        54        490.0      9.1      0.0          minor_axis = rp.minor_axis_length
   166                                                   # if area >= area_threshold and circ >= circularity_threshold:
   167        54         54.0      1.0      0.0          mask_info = {              
   168        54        173.0      3.2      0.0              'bbox': rp.bbox,
   169        54    1005597.0  18622.2     32.1              'contour': find_contours(seg, level=0.5)[0],
   170        54         23.0      0.4      0.0              'major_axis': major_axis,
   171        54          7.0      0.1      0.0              'minor_axis': minor_axis,
   172        54          7.0      0.1      0.0              'area': area,
   173        54        165.0      3.1      0.0              'radius': np.sqrt(area / np.pi),
   174        54         13.0      0.2      0.0              'circ': circ,
   175        54     105035.0   1945.1      3.4              'euler_number': rp.euler_number
   176                                                   }
   177                                           
   178        54      94765.0   1754.9      3.0          filtered_rows.iloc[idx] = pd.Series(mask_info)
   179                                           
   180         4       1647.0    411.8      0.1      df_filtered = filtered_rows[(filtered_rows["area"] > area_threshold) &
   181         2        337.0    168.5      0.0          (filtered_rows["circ"] >= circularity_threshold)]
   182         2          2.0      1.0      0.0      return df_filtered

  3.13 seconds - /Users/awitmer/Documents/LANL/2025/github_projects/ldrd_neat_ml/neat_ml/bubblesam/bubblesam.py:116 - analyze_and_filter_masks
```

</details>

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,330 @@
+def analyze_and_filter_masks(
+    masks_summary_df: pd.DataFrame,
+    area_threshold: float = 25,
+    circularity_threshold: float = 0.8
```
this has been updated to reflect the value used in the manuscript.

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,349 @@
+    output_path : str
+        File path to save the resulting figure.
+    """
+    fig, ax = plt.subplots(figsize=(10, 10))
```
could be `plot.subplots(1, 1)`

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,349 @@
+        )
+        ax.add_patch(rect)
+    ax.axis("off")
+    plt.savefig(output_path, bbox_inches='tight')
```
and here could be `fig.savefig(output_path, dpi=300, bbox_inches="tight")`

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,349 @@
+
+def process_image(
+    image_path: str,
+    output_dir: str,
```
consider unifying input/output paths to `Path` objects

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,349 @@
+        ), 
+        'wb'
+    ) as f:
+        pickle.dump(filtered_df, f)
```
with parquet, I think this could be, i.e.:

```python
filtered_df.to_parquet(output_dir / f'{image_basename}_masks_filtered.parqet.gz')
```

adamwitmer replied:
`parquet.gzip`

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,349 @@
+        plt.imshow(image)
+        show_anns(masks[1:])
+        plt.axis('off')
+        plt.savefig(os.path.join(
```
use `Path` object style formatting

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,349 @@
+        plot_filtered_masks(
+            original_image=image,
+            masks_summary_df=filtered_df,
+            output_path=os.path.join(
```
use `Path` object style formating

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,349 @@
+        Dataframe containing absolute image filepaths.
+        Requires 'image_filepath'.
+    output_dir : str | Path
+        Target directory for _masks_filtered.pkl + summary CSV.
```
will be `parquet` not `pickle`

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,256 @@
+
+_skip_unless_available()
+
+def test_setup_cuda_does_not_crash_on_cpu(monkeypatch):
```
remove `monkeypatch` testing

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,256 @@
+
+def test_load_image_missing_file_raises(tmp_path: Path) -> None:
+    missing_path = tmp_path / "does_not_exist.png"
+    expected = re.escape(f"Image at path {missing_path} not found.")
```
i dont think `re.escape` is necessary here...

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,256 @@
+    save_masks(masks, str(tmp_path))
+    out_file = tmp_path / "mask_0.png"
+
+    actual = cv2.imread(str(out_file), cv2.IMREAD_GRAYSCALE)
```
unnecessary `str` coercion here

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,256 @@
+    """
+    np.random.seed(0)
+    img_fp = image_with_circles_fixture
+    stem = Path(img_fp).stem
```
unnecesarry `Path` coercion

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,256 @@
+        "checkpoint_path": "./neat_ml/sam2/checkpoints/sam2_hiera_large.pt",
+        "device": "cpu",
+    }
+    df_in = pd.DataFrame({"image_filepath": [str(image_with_circles_fixture)]})
```
unnecessary `str` coercion if unifying to use `Path` objects.

ajwitmer committed:
MAINT: PR #3 revisions

* remove unnecessary type hints
* fix mypy errors
* fix tests
* update figure handles/expected images
* add gpu support for mps
* remove unnecessary type coercions
* unify file path handling
* update README.md
* remove unnecessary functions/tests
* remove monkeypatch testing

ajwitmer committed:
MAINT: fix conflicts after rebase

* fix testing logic
* update README.md
* fix logic for bubble detection

ajwitmer committed:
TST: remove leftover git lfs files

ajwitmer committed:
CI: fix github CI yaml file

ajwitmer committed:
MAINT, TST, CI: PR #3 revisions

* modify CI to download sam2 checkpoints for tests
* modify tests to use tiny sam2 model to improve speed
* adjust model weights filpath handling and update mypy type hints
* fix logic of sam2 _build_model to call only once
* update readme
* remove slow test skips
* remove unnecessary tests

ajwitmer committed:
DOC, TST, MAINT: PR #3 revisions

* update README.md with caveat for pytorch installations
* fix type hints
* add testing assertions for detected bubbles
* add torch, torchvision dependencies to pyproject.toml

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,256 @@
+    assert "area" in masks[0]
+    assert masks[0]["area"] ==  seg.sum()
+
+def test_run_bubblesam_cpu(tmp_path: Path, image_with_circles_fixture: Path):
```
from https://github.com/pytorch/pytorch/issues/84936, it cannot be assumed that results on `mps` and `cuda` will be numerically identical.

ajwitmer committed:
TST: update/add tests

* move fixtures to conftest
* remove bubblesam detection value assertions
* add tests for warnings/errors

ajwitmer committed:
TST: skip test intended for ``mps`` backend check

* add skip test when mps backend is not available

ajwitmer committed:
BUG: fix font download error

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,272 @@
+    assert isinstance(seg, np.ndarray)
+    assert seg.dtype is np.dtype("bool")
+    assert "area" in masks[0]
+    assert masks[0]["area"] ==  seg.sum()
```
extra space here, i.e. `==  seg.sum()`

adamwitmer commented on neat_ml/workflow/lib_workflow.py:
```diff
@@ -38,66 +39,98 @@ def get_path_structure(
     paths : dict[str, Path]
         Paths from get_path_structure() (proc_dir, det_dir if built).
-    
+        
```
whitespace error

adamwitmer replied:
unnecessary whitespace change

adamwitmer commented on neat_ml/workflow/lib_workflow.py:
```diff
@@ -106,7 +139,7 @@ def stage_detect(
     paths: dict[str, Path]
 ) -> Optional[pd.DataFrame]:
```
Modify this in line with https://github.com/lanl/ldrd_neat_ml/pull/2#discussion_r2739790068

tylerjereddy replied:
I'll mark it unresolved actually, since your explanation "I decided to leave this as is" is lacking in detail and effort, as discussed earlier today.

I do expect you to look at what other major projects do and use that as an example of what we should do, and/or to look at documentation around the matter. I'm consistently seeing things looking rushed, and not checking how this should be done is just another example of not putting in that effort.

tylerjereddy replied:
It sounds like this typing may be "ok" based on in person discussion because of a `None` propagation that happens through another called function.

That's probably the kind of explanation that would have been more helpful than "I decided to leave this as is," to save the reviewer a bit of time. But good that it may actually make sense, if somewhat complicated.

ajwitmer committed:
MAINT: PR #3 revisions

* remove unnecessary argument assignments
* improve docstrings
* minor style fixes
* remove global value assignment
* update README.md
* add inline explanation comments
* remove unnecessary test for directory handling
* remove unnecessary str coercions

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
As mentioned above, I modified this testing module to run the tests using the smallest version of `SAM2` (i.e. `sam2_hiera_tiny.pt`) and reduced mask settings to increase the speed of the test-suite. Locally on `ARM M2 Macbook` with torch `mps` gpu backend, the full test-suite runs in `01:47`; on `gp160` with torch `cuda` gpu backend the test-suite runs in `02:40`; In the CI (using only cpu) the test-suite passes in approximately `06:30` across various instances (whereas it was erroring out previously because of memory issues).

adamwitmer commented on neat_ml/tests/test_workflow.py:
I parametrized workflow tests across `opencv` and `bubblesam` detection methods to avoid duplication of tests that run the same module with different methods.

adamwitmer commented on neat_ml/workflow/lib_workflow.py:
Here I combined the modules `stage_opencv` and `stage_bubblesam` into a single function, `run_detection`, which programmatically performs the appropriate steps (e.g. preprocessing for `opencv`), and gathers the correct image dataset and save directory file paths for the given method.

ajwitmer committed:
DOC: update README.md

tylerjereddy commented:
@adamwitmer Ok, I note that this is being opened for review a week later than the deadline/schedule, and the programmatic polymer project is also behind schedule, so my review will likely be delayed accordingly.

Make sure you've really checked this carefully.

tylerjereddy commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,360 @@
+    mask_settings: dict[str, Any],
+    rng: Generator,
+    debug: bool = False,
+) -> pd.DataFrame:
```
Why does this function return a `pd.DataFrame` when it can optionally error out with `FileNotFoundError` below given your decision at https://github.com/lanl/ldrd_neat_ml/pull/3#discussion_r2784737984? I know which one I think is correct, but you've presented me with both return type options, so how can they both make sense? Is this carefully considered/presented?

ajwitmer committed:
MAINT: modify logic for return DataFrame from detection

* remove optional return from ``stage_detect`` in favor of
returning an empty dataframe if detection step skipped due
to any of multiple checks that throw log warnings and return
`None`
* modify logic to check for empty dataframe return in ``run_workflow``
* add edge case test for returning empty dataframe

tylerjereddy commented on README.md:
````diff
@@ -5,24 +5,54 @@
+After installation, the test-suite can be run using:
+
+```
+pyton -m pytest
````
This probably won't work unless you have really weird command aliasing.

Also, it would be helpful to keep PRs confined to changes actually needed for their scope. A small separate PR that handles documentation for running tests is probably much easier for me to review than having general testing docs squished into this large PR.

tylerjereddy commented on README.md:
```diff
@@ -5,24 +5,54 @@
+
+### Running the test-suite
+
+After installation, the test-suite can be run using:
```
Probably remove the hyphen

tylerjereddy commented on neat_ml/download_checkpoints.diff:
```diff
@@ -0,0 +1,80 @@
+-# Download each of the four checkpoints using wget
+-# echo "Downloading sam2_hiera_tiny.pt checkpoint..."
+-# $CMD $sam2_hiera_t_url || { echo "Failed to download checkpoint from $sam2_hiera_t_url"; exit 1; }
++Download each of the four checkpoints using wget
```
Why did you take the commented line `# Download each of the four checkpoints using wget` and remove the leading `#`?

tylerjereddy commented on .gitmodules:
```diff
@@ -0,0 +1,4 @@
+[submodule "neat_ml/sam2"]
+	path = neat_ml/sam2
+	url = https://github.com/facebookresearch/sam2.git
```
One thing we discussed providing in person, which I don't see here, is a clear justification for needing a submodule? Nidhin was never really able to clearly explain to me?

You can just pull in the model weights from Hugging Face as usual/standard as described at https://github.com/facebookresearch/sam2?tab=readme-ov-file#load-from--hugging-face, and then `pip install -e git+https://github.com....@hash` from the repo hash you need. I don't think we want to vendor code unless there's a really strong reason to do so (i.e., we don't vendor OpenCV, etc.).

tylerjereddy commented on .github/workflows/ci.yml:
```diff
@@ -16,19 +16,35 @@ jobs:
       run: |
-        python -m pip install -v ".[dev]"
+        python -m pip install -e ./neat_ml/sam2/
+        python -m pip install -v -e ".[dev]"
```
Why do we need an editable install of our own project in CI? We shouldn't need that. It could be "ok" to test it, but as a separate CI job if you really want that.

tylerjereddy commented on neat_ml/download_checkpoints.diff:
Feels hacky -- as my other comments note, the standard way of pulling in weights is to use Hugging Face as documented at https://github.com/facebookresearch/sam2?tab=readme-ov-file#load-from--hugging-face.

Which probably also removes the need for the submodule approach, which doesn't appear to be properly motivated as far as I can tell.

tylerjereddy reviewed (changes requested):
I did a first review pass looking at the need for vendoring SAM2 as a submodule, which is something I asked about in person and didn't receive a clear response on--maybe I'm missing the explanation for needing that here, but I don't see it.

Unless there's a strong/clear justification, then following the instructions to pull in weights via Hugging Face (https://github.com/facebookresearch/sam2?tab=readme-ov-file#load-from--hugging-face) and then `pip install -e ... github URL@hash` should suffice and avoid needing to vendor their entire project here.

I'll try to return for more code review later/tomorrow.

tylerjereddy commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,273 @@
+    try:
+        model.setup_cuda()
+    except Exception as e:
+        pytest.fail(f"setup_cuda() raised an unexpected exception on CPU: {e}")
```
1. I don't think you need this--if it errors out the test will fail naturally on a bare call of `model.setup_cuda()`, and the feedback from that should suffice.
2. If the testsuite is sufficiently thorough, and `setup_cuda()` is something that happens when doing inference on CPU anyway, shouldn't the testsuite already hit this problem in a CPU-only scenario when doing normal functional testing? May need a clearer motivation for why this test needs to exist.

tylerjereddy commented on neat_ml/tests/test_workflow.py:
```diff
@@ -178,3 +230,54 @@ def test_stage_detect_unknown_method_error(
+    tmp_path,
+    image_with_circles_fixture,
+    reduced_mask_settings,
+    mock_tiny_weights
```
The need to mock patch hack the `DEFAULT_MASK_SETTINGS` here looks like a design problem--you can just pass a custom mask dictionary to `run_bubblesam` as needed, so it seems like you should have the option to propagate the optional dictionary/ies via `run_detection`, rather than encouraging the user to physically edit our own library code (which is also what causes the "need" for mock patching here).

tylerjereddy commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,360 @@
+    output_dir: Path,
+    *,
+    model_cfg: dict[str, Any] = DEFAULT_MODEL_CFG,
+    mask_settings: dict[str, Any] = DEFAULT_MASK_SETTINGS,
```
These arguments are violating the "no mutable default arguments" principle. I'm not sure why `ruff` isn't picking it up, but if you change the arguments to `{}` it will correct detect it as shown below. Since we've discussed this before and added the settings to `ruff.toml` a few months ago, please check your functions carefully for this before presenting for review.

```
B006 Do not use mutable data structures for argument defaults
   --> neat_ml/bubblesam/bubblesam.py:306:37
    |
304 |     *,
305 |     model_cfg: dict[str, Any] = DEFAULT_MODEL_CFG,
306 |     mask_settings: dict[str, Any] = {},
    |                                     ^^
307 |     debug: bool = False,
308 | ) -> pd.DataFrame:
    |
help: Replace with `None`; initialize within function

Found 1 error.
```

tylerjereddy replied:
The reason for the false negative is described at: https://github.com/astral-sh/ruff/issues/21142. See also: https://github.com/astral-sh/ruff/pull/21189 for a confusing discussion. In short, best to avoid the mutable default args.

tylerjereddy commented on neat_ml/tests/test_workflow.py:
```diff
@@ -101,69 +165,57 @@ def test_stage_opencv_pipeline_runs(
+            shutil.copy(
+                mock_tiny_weights,
+                chkpt_dir,
+            )  # type: ignore[call-overload]
```
this looks pretty awkward; maybe will be a bit less messy with Hugging Face; i.e., `predictor = SAM2ImagePredictor.from_pretrained("facebook/sam2-hiera-tiny")` or similar.

tylerjereddy commented on neat_ml/tests/test_workflow.py:
```diff
@@ -178,3 +230,54 @@ def test_stage_detect_unknown_method_error(
+    }
+    paths = {"det_dir": det_dir}
+
+    df_out = wf.run_detection(ds, paths)
```
In addition to being able to pass the mask and checkpoint/weight settings via the "public API," it would also likely be useful if this test could be parametrized over available device backends (CPU, GPU, skipping when not avail) so we can see which cases/backends are passing explicitly.

tylerjereddy commented on neat_ml/workflow/lib_workflow.py:
```diff
@@ -38,75 +39,107 @@ def get_path_structure(
+    elif proc_dir.is_dir():
+        img_paths = chain(proc_dir.glob("**/*.tiff"),
+            proc_dir.glob("**/*.tif")
+        )  # type: ignore[assignment]
```
Just noticing that this seems overly complex, testsuite passes locally with simply `img_paths = proc_dir.glob("**/*.tif*")`.

tylerjereddy commented on neat_ml/workflow/lib_workflow.py:
```diff
@@ -38,75 +39,107 @@ def get_path_structure(
+    if method.lower() == "opencv":
+        df_out = run_opencv(df_imgs, det_dir, debug=debug)
+    else:
+        df_out = run_bubblesam(df_imgs, det_dir, debug=debug)
```
As noted elsewhere--almost certainly want a way to propagate mask, weight, CPU/GPU backend settings and so on to `run_bubblesam()` via the parent function here, so those can easily be controlled.

It is true that `DEFAULT_MODEL_CFG` seems to automatically pick the "right" device, but being able to set that manually seems important, and is also useful for parametrized testing over the different options.

tylerjereddy replied:
(and don't forget the issues with mutable default func arguments of course)

tylerjereddy commented on neat_ml/workflow/lib_workflow.py:
```diff
@@ -38,75 +39,107 @@ def get_path_structure(
+) -> pd.DataFrame:
     """
-    Route detection to OpenCV based on ``dataset_config`` method
+    Route detection to OpenCV or BubbleSAM based on dataset.method.
```
`method` still seems to be coming from `dataset_config` in the code below? why the docstring shim related to that?

tylerjereddy commented on neat_ml/workflow/lib_workflow.py:
```diff
@@ -117,14 +150,17 @@ def stage_detect(
-        df_out = stage_opencv(dataset_config, paths)
-        return df_out 
+    if method in ["opencv", "bubblesam"]:
+        df_out = run_detection(dataset_config, paths)
```
`stage_detect` and `run_detection` are both "public" functions at the moment. Probably need a way to pass CPU vs. GPU, mask settings, weights/checkpoint choices to them via any given possible public incantation. This also affects the usage of mock/patching in the testsuite (shouldn't be needed if you can pass those options through cleanly).

tylerjereddy commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,360 @@
+
+def analyze_and_filter_masks(
+    masks_summary_df: pd.DataFrame,
+    area_threshold: float = 25,
```
if this is a float, maybe make the `25` a float

tylerjereddy commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,360 @@
+    area_threshold : float
+        Minimum area for the mask to be retained. Value of minimum area was determined
+        by hand-tuning based on visual observation to exclude objects that were
+        too small to be considered bubbles in the microscopy images 
```
Are you referring to the "default value" here? Otherwise, the setting is up to the user.

tylerjereddy commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,360 @@
+    out_dir.mkdir(parents=True, exist_ok=True)
+
+    sam_model = SAMModel(**model_cfg)
+    sam_model.setup_cuda()
```
as noted elsewhere, I believe `setup_cuda()` ignores whether or not a user selected a CPU backend and may still warn about i.e., MPS?

tylerjereddy commented on neat_ml/data/bubblesam_detection_test.yaml:
```diff
@@ -0,0 +1,12 @@
+    
+    detection:
+      img_dir: 'path/to/pooch/images/images_raw.tiff'
+      debug: True
```
I used a modified version of this YAML file (below the fold)

<details>

```yaml
roots:
  work: /Users/treddy/rough_work/neat_ml

datasets:
  - id: tyler_test
    method: BubbleSAM
    class: boo
    time_label: foo
    
    detection:
      img_dir: '/Users/treddy/rough_work/neat_ml/image.tiff'
      debug: True
```

</details>

with a small image of bubbles converted to TIFF format (below the fold)

<details>

[image.tiff](https://github.com/user-attachments/files/25402401/image.tiff)

</details>

and then ran this on my ARM Mac laptop: `python run_workflow.py --config ~/rough_work/neat_ml/bubblesam_detection_test.yaml`

It seemed to be taking a really long time just for that one image. Perhaps as more productive feedback, it wasn't really clear to me how I could tune parameters to make things go faster/differently without having to edit our raw source code? Which is maybe not optimal for the end user trying this code on a different dataset that may need tweaking of inputs, like circularity thresholds and choice of weights, etc.

tylerjereddy replied:
Segmentation took about 12 minutes with the `yaml` file below the fold on this branch and using the small TIFF  image above on `gp160`. If I reduce to `crop_n_layers: 1`, the run time drops to 18 seconds and the masking still looks just fine. I found the `circularity_threshold: 0.92` to be fairly mediocre at distinguishing between squares and circles (bottom image), but turning it up does seem to move things in the right direction at least I suppose.


<details>

```yaml
roots:
  work: /home/treddy/LANL/gitlab_projects/ldrd_neat_ml/output_test

datasets:
  - id: tyler_testing
    method: BubbleSAM
    class: ''
    time_label: ''

    detection:
      img_dir: "/home/treddy/rough_work/neat_ml/march_9_2026/small_sample.tiff"
      debug: True
      area_threshold: 1.0
      circularity_threshold: 0.5
      model_cfg:
        mask_settings:
          points_per_side: 32
          points_per_batch: 128
          pred_iou_thresh: 0.80
          stability_score_thresh: 0.80
          stability_score_offset: 0.10
          crop_n_layers: 4
          box_nms_thresh: 0.10
          crop_n_points_downscale_factor: 1
          min_mask_region_area: 5
          use_m2m: True
        checkpoint_path: "facebook/sam2.1-hiera-large"
        device: "gpu"
```

</details>

<img width="1548" height="1051" alt="image" src="https://github.com/user-attachments/assets/e9ceaf3d-deb9-418f-a279-bccfbe996887" />

<img width="1516" height="1019" alt="image" src="https://github.com/user-attachments/assets/638c5c5e-f8c4-4852-a736-66b904e1f8a6" />

ajwitmer committed:
MAINT: address reviewer comments

* parameterize tests over cpu/gpu
* modify yaml input to pass default model args
* fix docstrings
* remove vendored sam2 module
* remove mock test params
* modify CI to use new sam2 module imports
* modify workflow to use updated sam2.1 model/weights
* remove unnecessary tests
* fix test assertions

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,273 @@
+        device="cuda",
+    )
+    model.setup_cuda()
+    if torch.cuda.get_device_properties(0).major >= 8:
```
I added a caveat to the associated lines in `SAMModel.setup_cuda()` from the `SAM2AutomaticMaskGenerator` jupyter notebook, where they reference https://docs.pytorch.org/docs/stable/notes/cuda.html#tensorfloat-32-tf32-on-ampere-and-later-devices:

> Starting in PyTorch 1.7, there is a new flag called allow_tf32. This flag defaults to True in PyTorch 1.7 to PyTorch 1.11, and False in PyTorch 1.12 and later. This flag controls whether PyTorch is allowed to use the TensorFloat32 (TF32) tensor cores, available on NVIDIA GPUs since Ampere, internally to compute matmul (matrix multiplies and batched matrix multiplies) and convolutions.
>
>TF32 tensor cores are designed to achieve better performance on matmul and convolutions on torch.float32 tensors by rounding input data to have 10 bits of mantissa, and accumulating results with FP32 precision, maintaining FP32 dynamic range.

This includes `NVIDIA` gpus from the RTX 30 series onwards, and doesnt need to be set for older GPU's

adamwitmer commented on neat_ml/tests/test_bubblesam.py:
```diff
@@ -0,0 +1,273 @@
+def test_bubblesam_detection_generates_pngs_cpu(
+    tmp_path: Path,
+    image_with_circles_fixture: Path,
+    real_sam_model: SAMModel,
```
I modified this test to parameterize over `cuda` and `mps` devices.

tylerjereddy commented on neat_ml/download_checkpoints.diff:
```diff
@@ -0,0 +1,80 @@
+diff --git a/checkpoints/download_ckpts.sh b/checkpoints/download_ckpts.sh
```
> because the test suite still passes using the newer checkpoints.

In principle, it should probably fail using the older checkpoints if the tests are sufficiently stringent to require the improved model performance.

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,360 @@
+    """
+    if len(anns) == 0:
+        return
+    sorted_anns = sorted(anns, key=(lambda x: x['area']), reverse=True)
```
I modified the test associated with `show_anns` to enforce the criteria that "smaller" annotations are plotted on top of "larger" annotations so that they are not covered up.

tylerjereddy commented on README.md:
````diff
@@ -70,34 +60,58 @@ datasets:
-```bash
-module load cudatoolkit/24.7_12.5  
-```  
-
````
Let's get proper instructions for running on Chicoma--you've had a few months to reproduce the paper results there now, so I think that's fair to ask for. And you need to run there again for Mihee soon on block copolymer images.

ajwitmer committed:
MAINT: PR #3 revisions

* fix docs
* modify handling of device setting
* modify tests to handle device settings appropriately
* add in-line comments clarifying design choices

tylerjereddy commented on README.md:
````diff
@@ -42,7 +72,23 @@ datasets:
+### To run the code in CHICOMA:  
+```bash
+module load cudatoolkit/24.7_12.5  
+```  
````
The 6 hour and 20 minute inference runtimes on M-series and 4090 respectively do seem a bit suspicious based on upstream discussions like https://github.com/facebookresearch/sam2/issues/543.

tylerjereddy replied:
Ok, I already asked in person if there was any progress on that months-old issue and there wasn't.

I'm not sure the issue really even captures some of the important considerations. If the ROC AUC can be preserved with less ridiculous settings, and/or more appropriate image preprocessing, that seems like something that might have made sense to investigate and understand over the past several months so that you're well positioned to iterate on current results and respond to new image data with reasonable turnaround times.

tylerjereddy replied:
> the most computationally expensive tasks are skimage.measure image processing functions

What 3 things have you tried to deal with the bottlenecks over the past months? The bubble processing is arguably one of the most important components of the work, right?

1. OpenCV tends to have faster alternatives to many `skimage` functions (may require shims to get exact same results) written in C++ for CPU and CUDA for GPU when available via i.e., `.cuda.func` calls or something like that.
2. https://github.com/rapidsai/cucim has a GPU-based implementation of some of the `skimage` API via the `cucim.skimage` module. (may require shims to get exact same results and deal with CPU/GPU device transfers efficiently)
3. Can check for looping that might be vectorized (for example, CPU/GPU device transfers may be expensive but justifiable if you "batch" enough data to make the GPU transfer worth it)
4. Better ideas than mentor suggestions, etc.

ajwitmer committed:
MAINT, TST, DOC: PR #3 revisions

* modify handling of model_cfg in yaml file
* add ValueError check for model_cfg input
* add test for ValueError raise
* update docs to reflect changes

ajwitmer committed:
MAINT, DOC: PR #3 revisions

* update bubblesam yaml file to reflect change in detection_cfg
* update instructions for running workflow on chicoma

adamwitmer commented on neat_ml/bubblesam/SAM.py:
```diff
@@ -0,0 +1,127 @@
+        stored at:
+
+        https://github.com/facebookresearch/sam2/blob/main/
+        notebooks/automatic_mask_generator_example.ipynb
```
need to add `Parameters` and `Returns` docstring here

adamwitmer commented on neat_ml/workflow/lib_workflow.py:
```diff
@@ -117,14 +147,17 @@ def stage_detect(
-    df_out: Optional[pd.DataFrame]
+    df_out: pd.DataFrame
         dataframe containing summary of opencv bubble detection
         information 
```
OR empty dataframe that propagates through `run_workflow` and raises a warning

adamwitmer commented on mypy.ini:
```diff
@@ -1,5 +1,5 @@
 [mypy]
-exclude=joblib_cache
+exclude=joblib_cache|^neat_ml/sam2/.*
```
we're not vendoring the `sam2` module anymore

adamwitmer commented on ruff.toml:
```diff
@@ -1,2 +1,4 @@
+exclude = ["neat_ml/sam2"]
+
```
not vendoring `sam2` anymore

ajwitmer committed:
MAINT, DOC: PR #3 revisions

* remove leftover settings from vendoring sam2
* update README.md with flavored markdowns
* add/fix inline comments and docstrings

ajwitmer committed:
MAINT: revert change to mypy missing imports

ajwitmer committed:
MAINT, ENH, DOC, TST: PR #3 revisions

* add cuda enabled image post-processing with `cucim`
* add instructions for installing cucim on cuda enabled systems
* modify contour finding to use opencv
* fix test to enforce assertions for checking sam2.1 detections
* update mypy types

ajwitmer committed:
MAINT, TST, DOC: PR #3 revisions

* fix parquet data handling/saving/loading
* add test to check that parquet logic behaves as expected
* modify pyproject to install fastparquet instead of pyarrow
* revert unnecessary change

ajwitmer committed:
BLD: revert change removing pyarrow in pyproject.toml

adamwitmer commented on neat_ml/tests/test_workflow.py:
```diff
@@ -178,3 +230,54 @@ def test_stage_detect_unknown_method_error(
+    """
+    det_dir = tmp_path / "det"
+    det_dir.mkdir(parents=True)
+    ds = {"id": "BS3", "method": "BubbleSAM", "detection": {"img_dir": "bad/path"}}
```
It is not necessary to parameterize this test over both methods because this error is only raised when using the bubblesam method. If using opencv, a similar error is raised by the function `neat_ml.opencv.preprocessing.process_directory`, i.e.:

https://github.com/lanl/ldrd_neat_ml/blob/02e90947486506c5a327cdad067068d722bfee38/neat_ml/opencv/preprocessing.py#L110-L112

adamwitmer commented on README.md:
```diff
@@ -8,50 +8,128 @@ Download the package from the following GitHub repository:
+> specific GPU on the users system. For download instructions and information visit: 
+> https://pytorch.org/get-started/locally/
+>
+> Measurement of detected bubbles using `SAM2` is optionally handled using the `cucim.skimage.measure` library.
```
`of bubbles detected by SAM2`

adamwitmer commented on README.md:
```diff
@@ -8,50 +8,128 @@ Download the package from the following GitHub repository:
              of raw image next to bounding box overlay.)
+      # below only include when using the ``BubbleSAM`` detection method
+      area_threshold: threshold for minimum area of a detected bubble
+      circularity_threshold: threshold for minimum circularity of a detected bubble
```
(float)

ajwitmer committed:
DOC, TYP: PR #3 revisions

* update README for clarity
* remove unnecessary None type return in test signature

ajwitmer committed:
TYP: remove unnecessary `None` return typing in tests

tylerjereddy commented on README.md:
```diff
@@ -8,50 +8,128 @@ Download the package from the following GitHub repository:
       img_dir: path/to/image/data (Can be a directory of ``.tiff`` images or a path to a single ``.tiff`` image.)
       debug: True/False for debug (`True` will save side-by-side figure
              of raw image next to bounding box overlay.)
+      # below only include when using the ``BubbleSAM`` detection method
```
```suggestion
      # only include the content below when using the ``BubbleSAM`` detection method
```

or similar

tylerjereddy commented on README.md:
```diff
@@ -8,50 +8,128 @@ Download the package from the following GitHub repository:
              of raw image next to bounding box overlay.)
+      # below only include when using the ``BubbleSAM`` detection method
+      area_threshold: (float) threshold for minimum area of a detected bubble
+      circularity_threshold: (float) threshold for minimum circularity of a detected bubble
```
One thing I'm wondering about as I read through is the asymmetry in allowing the user to control SAM2 settings via input deck, but not OpenCV -- if I'm not mistaken OpenCV params are controlled via editing the source code in `neat_ml/opencv/detection.py`.

I don't think you'll have time to remedy that for now, but opening an issue to track it may be sensible. For example,  `params.minCircularity` for OpenCV obviously has a parallel to circularity settings specified for SAM2 in the input deck/YAML file.

adamwitmer replied:
I opened issue #27 to this effect.

ajwitmer committed:
MAINT: address reviewer comments

* move sam2 install to pyproject dependencies
* update README
* add image metric calculation to joblib cache memory
* update CI
* fix test skip logic
* update pyproject.toml

adamwitmer commented on neat_ml/bubblesam/bubblesam.py:
```diff
@@ -0,0 +1,384 @@
+
+from .SAM import SAMModel
+
+memory = joblib.Memory("joblib_cache", verbose=0)
```
I added a `@memory.cache` decorator to `analyze_and_filter_masks` because it is a computationally expensive task that may be recomputed multiple times.

ajwitmer committed:
TST, BUG: fix test skip assertion

tylerjereddy commented:
ok, in it goes, thanks
